.github/workflows/e2e_test.yaml

```diff
@@ -0,0 +1,191 @@
+      - name: Generate Run ID
+        id: run-id
+        run: |
+          echo "run-id=e2e-$(LC_ALL=C tr -dc 'a-z' < /dev/urandom | head -c8)" >> $GITHUB_OUTPUT
+
+  deploy-e2e-test-runner:
+    name: Deploy End-to-End Test Runner
+    runs-on: ubuntu-latest
+    needs: [ build-charm, run-id ]
+    steps:
+      - name: Setup Lxd Juju Controller
+        uses: charmed-kubernetes/actions-operator@main
+        with:
+          juju-channel: 3.1/stable
+          provider: lxd
+
+      - name: Install GitHub Cli
+        run: which gh || sudo apt install gh -y
+
+      - name: Create Testing Juju Model
+        run: juju add-model testing
+
+      - name: Set Testing Model Proxy Configuration
+        run: |
+          juju model-config juju-http-proxy=$http_proxy
+          juju model-config juju-https-proxy=$https_proxy
+          juju model-config juju-no-proxy=$no_proxy
+
+      - name: Change Testing Model Logging Level
+        run: juju model-config logging-config="<root>=INFO;unit=DEBUG"
+
+      - name: Download github-runner Charm
+        uses: actions/download-artifact@v3
+        with:
+          name: dangerous-test-only-github-runner_ubuntu-22.04-amd64.charm
+
+      - name: Deploy github-runner Charm
+        run: |
+          cp github-runner_ubuntu-22.04-amd64.charm /home/$USER/github-runner_ubuntu-22.04-amd64.charm
+          juju deploy /home/$USER/github-runner_ubuntu-22.04-amd64.charm \
+            ${{ needs.run-id.outputs.run-id }} \
+            --base ubuntu@22.04 \
+            --config path=${{ secrets.E2E_TESTING_REPO }} \
+            --config token=${{ secrets.E2E_TESTING_TOKEN }} \
+            --config virtual-machines=1 \
+            --config test-mode=insecure
+
+      - name: Watch github-runner
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        timeout-minutes: 30
+        run: |
+          juju debug-log --replay --tail &
+          
+          while :; do
+            JOBS=$(gh api \
+            -H "Accept: application/vnd.github+json" \
+            -H "X-GitHub-Api-Version: 2022-11-28" \
+            /repos/${{ secrets.E2E_TESTING_REPO }}/actions/runs/$GITHUB_RUN_ID/attempts/$GITHUB_RUN_ATTEMPT/jobs)
+            CONCLUSION=$(echo $JOBS | jq -r '.jobs[] | select(.name == "End-to-End Test") | .conclusion')
+            STATUS=$(echo $JOBS | jq -r '.jobs[] | select(.name == "End-to-End Test") | .status')
+            if [[ $STATUS != "queued" && $STATUS != "in_progress" ]]; then
+              break
+            fi
+            sleep 10
+          done
+          if [[ $STATUS != "completed" || $CONCLUSION != "success" ]]; then
+            echo "test workflow failed with status: $STATUS, conclusion: $CONCLUSION"
+            kill $(jobs -p)
+            exit 1
+          fi
+
+  e2e-test:
+    name: End-to-End Test
+    needs: [ build-charm, run-id ]
+    runs-on: [ self-hosted, linux, x64, "${{ needs.run-id.outputs.run-id }}" ]
+    steps:
+      - name: Echo hello world
```

Comment:
Why is this echo here?

Reply:
This is how the runners are tested, so this is a fairly basic test where the runner only has to be able to connect and be able to echo

Reply:
Ok, if possible would be nice to have a comment clarifying that 👍

Reply:
Added, thanks!